.github/workflows/linux.yml

```
@@ -150,3 +150,55 @@ jobs:
     - name: build
       run: |
         make -j 2
+
+
+  linux-debug-intel-oneapi:
+    # parallel debug build with Intel oneAPI including MPI and MKL
+
+    name: linux debug intel oneapi
+    runs-on: [ubuntu-18.04]
+
+    steps:
+    - name: modules
+      run: |
+        # Based on https://github.com/oneapi-src/oneapi-ci
+        # Setup Intel repository
+        # oneapi-ci/scripts/setup_apt_repo_linux.sh
+        wget https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        sudo apt-key add GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        echo "deb https://apt.repos.intel.com/oneapi all main" | sudo tee /etc/apt/sources.list.d/oneAPI.list
+        sudo apt-get update -o Dir::Etc::sourcelist="sources.list.d/oneAPI.list" -o APT::Get::List-Cleanup="0"
+        # Install Intel packages -- see https://oneapi-src.github.io/oneapi-ci/#linux-apt
+        # oneapi-ci/scripts/install_linux_apt.sh
+        sudo apt-get install -y intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic \
+                                intel-oneapi-mpi-devel \
+                                intel-oneapi-mkl-devel
+        sudo apt-get clean
+    - name: info
+      run: |
+        source /opt/intel/oneapi/setvars.sh
+        mpiicpc -v
+        cmake --version
+    - name: checkout
+      uses: actions/checkout@v2
+    - name: configure
+      run: |
+        source /opt/intel/oneapi/setvars.sh
+        cmake -D CMAKE_BUILD_TYPE=Debug \
+              -D CMAKE_CXX_COMPILER=mpiicpc \
```

```suggestion
              -D CMAKE_CXX_COMPILER=mpiicpc \
```
Setting MPI compiler wrapper as CMAKE_CXX_COMPILER is not recommended. Is there a good reason for doing that here instead of choosing `icpc`?

---

I tried to split this line into
```
 -D CMAKE_CXX_COMPILER=icpc \
 -D MPI_CXX_COMPILER=mpiicpc \
```
or just use
```
 -D CMAKE_CXX_COMPILER=icpc \
```
which both causes a CMake error
```
-- Include /home/marc/dealii/cmake/setup_finalize.cmake
-- Performing Test DEAL_II_HAVE_USABLE_FLAGS_DEBUG
CMake Warning (dev) at /home/marc/dealii/build/CMakeFiles/CMakeTmp/CMakeLists.txt:16:
  Syntax Warning in cmake code at column 147

  Argument not separated from preceding token by whitespace.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at /home/marc/dealii/build/CMakeFiles/CMakeTmp/CMakeLists.txt:15 (add_executable):
  Target "cmTC_24f4d" links to item "-shared-intel -qopenmp -rdynamic
  -fuse-ld=gold -lpthread -Xlinker --enable-new-dtags -Xlinker -rpath
  -Xlinker " which has leading or trailing whitespace.  This is now an error
  according to policy CMP0004.


CMake Error at /opt/cmake/share/cmake-3.20/Modules/Internal/CheckSourceCompiles.cmake:89 (try_compile):
  Failed to generate test project build system.
Call Stack (most recent call first):
  /opt/cmake/share/cmake-3.20/Modules/CheckCXXSourceCompiles.cmake:76 (cmake_check_source_compiles)
  cmake/macros/macro_check_compiler_setup.cmake:73 (CHECK_CXX_SOURCE_COMPILES)
  cmake/setup_finalize.cmake:89 (CHECK_COMPILER_SETUP)
  cmake/setup_finalize.cmake:114 (_check_linker_flags)
  cmake/macros/macro_verbose_include.cmake:19 (INCLUDE)
  CMakeLists.txt:129 (VERBOSE_INCLUDE)


-- Configuring incomplete, errors occurred!
See also "/home/marc/dealii/build/CMakeFiles/CMakeOutput.log".
See also "/home/marc/dealii/build/CMakeFiles/CMakeError.log".
```

I don't know how to fix this. This failing test is listed in neither `CMakeOutput.log` nor `CMakeError.log`, and my `CMakeTmp` folder is empty after configuration fails.

---

Also, we still recommend passing the MPI wrappers via `CMAKE_{LANG}_COMPILER`:
https://github.com/dealii/dealii/blob/dce1df1fd0d0ad650232577cc79f8c2739fe5d96/cmake/configure/configure_1_mpi.cmake#L63-L78

---

Yeah. We should update this error message.

---

I tried to look a little bit more into the linker flags in https://github.com/masterleinad/dealii/runs/3032612756?check_suite_focus=true and see
```
[...]
-- Initial linker_flags: X  -shared-intel -qopenmp -rdynamic -fuse-ld=gold -lpthread -Xlinker --enable-new-dtags -Xlinker -rpath -Xlinker "/opt/intel/oneapi/mpi/2021.3.0/lib/release" -Xlinker -rpath -Xlinker "/opt/intel/oneapi/mpi/2021.3.0/lib"  X
-- Stripped linker_flags: X -shared-intel -qopenmp -rdynamic -fuse-ld=gold -lpthread -Xlinker --enable-new-dtags -Xlinker -rpath -Xlinker "/opt/intel/oneapi/mpi/2021.3.0/lib/release" -Xlinker -rpath -Xlinker "/opt/intel/oneapi/mpi/2021.3.0/lib" X
-- Performing Test DEAL_II_HAVE_USABLE_FLAGS_DEBUG
CMake Warning (dev) at /home/runner/work/dealii/dealii/CMakeFiles/CMakeTmp/CMakeLists.txt:16:
  Syntax Warning in cmake code at column 147

  Argument not separated from preceding token by whitespace.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at /home/runner/work/dealii/dealii/CMakeFiles/CMakeTmp/CMakeLists.txt:15 (add_executable):
  Target "cmTC_589a0" links to item "-shared-intel -qopenmp -rdynamic
  -fuse-ld=gold -lpthread -Xlinker --enable-new-dtags -Xlinker -rpath
  -Xlinker " which has leading or trailing whitespace.  This is now an error
  according to policy CMP0004.
[...]
```
note that `"/opt/intel/oneapi/mpi/2021.3.0/lib/release"` and `"/opt/intel/oneapi/mpi/2021.3.0/lib"` have been stripped somehwere.

---

If I print `CMAKE_REQUIRED_LIBRARIES` right before `CHECK_CXX_SOURCE_COMPILES`, it still contains the `"/opt/intel/oneapi/..."` paths.

Could this be a bug in `CHECK_CXX_SOURCE_COMPILES`? I can try to configure it with the `cmake 3.21` release candidate.

---

Updating `cmake` doesn't help. Same error message, but now highlighted in fancy colors. Also, BLAS and LAPACK will no longer be found.

---

`CMAKE_REQUIRED_LIBRARIES` for
- `mpiicpc`:
```
-shared-intel -qopenmp -rdynamic -fuse-ld=gold -lpthread;/opt/intel/oneapi/mpi/2021.3.0/lib/release/libmpi.so;pthread;dl;rt
```
- `icpc`:
```
-shared-intel -qopenmp -rdynamic -fuse-ld=gold -lpthread -Xlinker --enable-new-dtags -Xlinker -rpath -Xlinker "/opt/intel/oneapi/mpi/2021.3.0/lib/release" -Xlinker -rpath -Xlinker "/opt/intel/oneapi/mpi/2021.3.0/lib";/opt/intel/oneapi/mpi/2021.3.0/lib/libmpicxx.so;/opt/intel/oneapi/mpi/2021.3.0/lib/libmpifort.so;/opt/intel/oneapi/mpi/2021.3.0/lib/release/libmpi.so;pthread;dl;rt
```

Notice how the error message of `cmake` is truncated after the first quotation marks. Could it be that `cmake` or rather `CHECK_CXX_SOURCE_COMPILES` simply do not like quotation marks?

---

It works if I change all quotation marks like this right before calling `CHECK_CXX_SOURCE_COMPILES` like this:
```
STRING(REPLACE "\"" "'" CMAKE_REQUIRED_LIBRARIES "${CMAKE_REQUIRED_LIBRARIES}")
```
Compilation seems to succeed.

This doesn't seem right?! I'll try to come up with a minimal example and report it to the cmake group...

---

Issues reported with failing examples to the `CMake` developers.
- mpi: https://gitlab.kitware.com/cmake/cmake/-/issues/22402
- lapack: https://gitlab.kitware.com/cmake/cmake/-/issues/22403

---

We should probably convert to a semicolon-separated list...

---

Possibly using https://cmake.org/cmake/help/latest/command/separate_arguments.html.

---

I pushed some commits that should fix it. Feel free to ignore or propose to fix in a separate pull request.

---

Thank you for the patch! It's exactly what has been suggested in the issue I opened in the CMake repository.

It's interesting to see that this seems to break the `cuda` (issue with bundled boost) and `simplex` (missing references to opal) workflows.

Maybe specifying compilers and mpi wrappers will help. I'll give it a try.